{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[]},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"code","source":["#Visualizing the sample fastq.\n","!zcat < GLDS-280_Amplicon_SRR8985886_1.fastq.gz | head\n"],"metadata":{"id":"XIBQXnV-TTQD","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1680418176749,"user_tz":420,"elapsed":194,"user":{"displayName":"Ananya Kohli","userId":"06341108659029638783"}},"outputId":"d103f204-44e2-44e6-f9ff-5b30f3bba3ec"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["@SRR8985886.1 1/1\n","CCTACGGGGGGCTGCAGTGAGGAATCTTCCACAATGGGCGAAAGCCTGATGGAGCAACGCCGCGTGAAGGATGAAGGCCTTCGGGTTGTAAACTTCTTTTATGAGTGAAGAATATGACGGTAACTCATGAATAAGCACCGGCTAACTACGTGCCAGCAGCCGCGGTCATACGTAGGGTGCAAGCATTATCCGGAGTGACTGGGCGTAAAGAGTTGCGTAGGCGGTTTAATAGGTGAATAGTGAACCCTGGTGGCTCAACCATACAGATTATTATTCAAACCGTTAAACCAGAGAAGGGTAG\n","+\n","88BCCGGGFAA=188::::CFFEFGGGGG<<FEFFFGGGGGFGGGGGGFDFF8FF7EFG:CFGGGGGFGGGDGFFFGCGGGGGGGGGGGG9@FGGFFGGGFCGECFG?CFEFGFGFGGEC<<FGGGGGCF,C;9FGGGGGGGG7<BFCCFAFFG:CFCF<<CEEGGGCC?CFEEGGGGGE6+3C;CC@FE98C85E83+<<CCEGG=E*?7AF77@C:*:E,:CCEGEFFG7FCGD4=5:AF4>+:(,A**4C)03*/70)27*/0*+200<6@F76F))**2*;::*()1)*(+2;E>+:\n","@SRR8985886.2 2/1\n","CCTACGGGGGGCTGCAGTGAGGAATCTTCCACAATGGGCGAAAGCCTGATGGAGCAACGCCGCGTGAAGGATGAAGGCCTTCGGGTTGTAAACTTCTTTTATGAGTGAAGAATATGACGGTAACTCATGAATAAGCACCGGCTAACTACGTGCCAGCAGCCGCGGTCATACGTAGGGTGCAAGCATTATCCGGAGTGACTGGGCGTAAAGAGTTGCGTAGGCGGTTTAATAAGGGAATAGTGAAACCTGGTGGCTCAACCATAAAGCCTATTCTTCAAATTGTTAAACTCGAGCATGGTAG\n","+\n","8ACCCGGGAA58188:::::FFGGGGGGGGGGGGGGGGGGDFGGCGEFFFFGGGGGGGGGGEFGFGGFGGFDEFFFGFECFFFEGF:FGGFGGG9FFGFGGGGGGGFF@FFFFGGGGGGGGGGGGFGGGGFGGGGGCFFFGG7*8CFCGDFGGGFGGF:BC8:/:::/AE,:<FFEGFEGGGGGGG7@?FC>:CE;8:CDFGC>>:?E3>CFGGFFG:::>,:8::*,<40>C+077)<FA7*C::3(2A9(7).+90<?BGG999**+269)9+*/04)30,<<C77<*107(*0+;2++\n","@SRR8985886.3 3/1\n","CCTACGGGAGGCAGCAGTAGGGAATCTTCCGCAATGGACGCAAGTCTGACGGAGCAACGCCGCGTGAGTGAAGAAGGTTTTCGGATCGTAAAACTCTGTTGTTAGAGAAGCACAAGTGCTAGAGTAACTGTTAGCGCCTTGACGGTATCTAACCAGAAAGCCACGGCTAACTACGTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTGTCCGGATTTATTGGGCGTAAAGCGAGCGCAGGCGGTTCCTTAAGTCTGATGTGAAAGCCCCCGGCTCAACCGGGGAGGGTCATTGGAA\n"]}]},{"cell_type":"code","source":[],"metadata":{"id":"ERoYh4pkUxUl"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["!zcat < GLDS-280_Amplicon_SRR8985886_2.fastq.gz | head"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"pLSej23aucjI","executionInfo":{"status":"ok","timestamp":1680418890967,"user_tz":420,"elapsed":443,"user":{"displayName":"Ananya Kohli","userId":"06341108659029638783"}},"outputId":"454b6860-f027-4db0-ac6b-eafc345bca36"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["@SRR8985886.1 1/2\n","NACTACAGGGGTATCTAATCCGGTTCGCTCCCCACGCTTTCGTGCCTTAGCGTCAGAAATGGCCCAGTAACCTGCCTACGCCATCGGTGTTCCTTCTAATATCTACGGATTTCACTCCTACACTAGAAATTCCAGTTACCTCTACCATTCTCGAGTTTACCAGTTTGAATAATAGTCTGTCTGGTTGAGCCACCAGGTTTCACTATTCCCTTATTAACCCGCCGACGCACCTCTTCACGCCCGATCATTCTGTATAATGCTTGTCCTCTTGGTTATACCTGCTGTGCTGCGCCGGTCTGAT\n","+\n","#8ABADCFGGGGEGGGFGGGGGGGGGFCCE@DEFFCEEGGGGGFGFG<FGFGFDGG<FFEFDF<<@FGGGFFFGGGGEFCF7FEGCF8?FFG;9=,E?EF,2A@,<AFCCEDADEEE=*49ABB<,5,5CFFFBE9D9D94DD+D,+83@DF6DCF==,,CDFDCA,:,CF+A@,22=+C,,3BF<6,>8=,53,,+:@7++5C,>F7+*0A:BFEF,4,**))))/((*(0401*10*:3((((.())3+++++)+)2313)()6))))-)3,.).4))),.).()-))-(-31((((.)\n","@SRR8985886.2 2/2\n","NACTACAGGGGTATCTAATCCGGTTCGCTCCCCACGCTTTCGTGCCTTAGCGTCAGAAATGGCCCAGTAACCTGCCTACGCCATCGGTGTTCCTTCTAATATCTACGGATTTCACTCCTACACAAGAAATTCCAGTTACCTCTACCATTCTCGAGTTTAACAGTTTGAATAATAGTCTGTATGGTTGAGCCACCAGGTTTCACTATTCACTTATTAAACCTCCTACGCAACTCTTTACGAACAGTCACTCCGGATACTGCTTGCACCGTAAGGATGACCGCTGCTTCCGGCACGTAGGTCG\n","+\n","#8AA@D,CFGGGCFGGFFGGCFGGGGGGGGEFFDFGGGGGGGFFFGGGGGGGGGGGFGGGGGFEFGGGGGFGEFGGFFFGEGGCFEGFCBCFGGFECEFGDFGGG<DCECFG9EEEEGEFDFF,:,,:EEGGGGADDF?C>ED7=BD<AEEFD8BEGFDBF=DDEF9?,:F8AFE2;D=F<FGD<DE,>=<DFB<DE<,48CCFE79;=BEF*;,?,,;D,;58564=>>))4187:<***(0(1.*1*+))*3*(+)+++27=E),(,((,,/..().(-(((,)))((-((43,((.),\n","@SRR8985886.3 3/2\n","NACTACTGGGGTATCTAATCCTGTTTGCTACCCACGCTTTCGAGCCTCAGCGTCAGTTACAGACCAGAGAGTCGCCTTCGCCACTGGTGTTCCTCCCTATATCTACGCATTTCACCGCTACACATGGAATTCCACTCTCCTCTTCTGCACTCAAGTTCCCCAGTTTCCCATGACCCTCCCCCGTTGAGCCGGGGGCTTTCACCTCAGACTGAAGGAACCCCCCTCGCTCGTCTTACGCCCAATAAATCCGGCTAACGCTTGCCACCTCACGACTACCTCTGCTTCCTGCCGCTCGCTTGC\n"]}]},{"cell_type":"markdown","source":["#Oral Microbiome changes associated with space travel and evidence of anxiety\n","\n","The following code was used to analyze the oral microbiome data through saliva samples of 10 astronauts at three time points: pre flight, during flight and post flight. 16S rRNA was sequenced using Amplicon sequencing. Quality control was performed using Qiime2 plugin of DADA2.\n"],"metadata":{"id":"PFya0SYIJmGw"}},{"cell_type":"code","source":["#Loading the metadata file.\n","!cat space_metadata.tsv | cut -f2,10-13 > space_tmp.tsv"],"metadata":{"id":"-z-bIDEdJ28J","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1680468887170,"user_tz":420,"elapsed":264,"user":{"displayName":"Ananya Kohli","userId":"06341108659029638783"}},"outputId":"65dc58dd-303c-4b86-c586-e74eb849be94"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["ï»¿Source Name\tSample Name\tCharacteristics: Biosample\tCharacteristics: Organism\tCharacteristics: host\tProtocol REF\tParameter Value: Subject ID\tMaterial Type\tFactor Value: Spaceflight\tFactor Value: Time\tParameter Value: Sample Collection Timepoint\tParameter Value: Sample Storage Temperature\n","L-240\tSRR8985933\tSAMN11515255\tcellular organisms\tHomo sapiens\tsample collection\t1\tSaliva\tSpace Flight\tpre-flight\t240 days before launch\t-80 degree Celsius\n","L-240\tSRR8985934\tSAMN11515256\tcellular organisms\tHomo sapiens\tsample collection\t1\tSaliva\tSpace Flight\tpre-flight\t240 days before launch\t-80 degree Celsius\n","L-180\tSRR8985931\tSAMN11515257\tcellular organisms\tHomo sapiens\tsample collection\t1\tSaliva\tSpace Flight\tpre-flight\t180 days before launch\t-80 degree Celsius\n","L-90\tSRR8985932\tSAMN11515258\tcellular organisms\tHomo sapiens\tsample collection\t1\tSaliva\tSpace Flight\tpre-flight\t90 days before launch\t-80 degree Celsius\n","L-60\tSRR8985929\tSAMN11515259\tcellular organisms\tHomo sapiens\tsample collection\t1\tSaliva\tSpace Flight\tpre-flight\t60 days before launch\t-80 degree Celsius\n","FD7\tSRR8985930\tSAMN11515260\tcellular organisms\tHomo sapiens\tsample collection\t1\tSaliva\tSpace Flight\tduring flight\t7 days on ISS\t-80 degree Celsius\n","FD90\tSRR8985927\tSAMN11515261\tcellular organisms\tHomo sapiens\tsample collection\t1\tSaliva\tSpace Flight\tduring flight\t90 days on ISS\t-80 degree Celsius\n","R-14\tSRR8985928\tSAMN11515262\tcellular organisms\tHomo sapiens\tsample collection\t1\tSaliva\tSpace Flight\tduring flight\t14 days before return\t-80 degree Celsius\n","R-1\tSRR8985935\tSAMN11515263\tcellular organisms\tHomo sapiens\tsample collection\t1\tSaliva\tSpace Flight\tduring flight\t1 day before return\t-80 degree Celsius\n"]}]},{"cell_type":"code","source":["#Making the metadate file qiime2 format compatible.\n","!sed -i 's\\Sample Name\\sample-id\\g' space_tmp.tsv"],"metadata":{"id":"xOEiE3h7oVKI"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["#Visualizing the final metadata file.\n","!head space_tmp.tsv"],"metadata":{"id":"JHkvUz7coYkt"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Paired end sequencing was conducted. A manifest file for the paired end data was first created."],"metadata":{"id":"vvCgi73OJ9hq"}},{"cell_type":"code","source":["#Creating the manifest file.\n","#Loading the assay csv file.\n","\n","!tr ',' '\\t' < OSD-280-assays.csv > assay.tsv\n","!cat assay.tsv | cut -f1,14,15 > manifest_tmp.tsv\n","!sed -i 's\\Sample Name\\sample-id\\g' manifest_tmp.tsv\n","!sed -i 's\\Protocol REF\\forward-absolute-filepath\\g' manifest_tmp.tsv\n","!sed -i 's\\Parameter Value: README\\reverse-absolute-filepath\\g' manifest_tmp.tsv\n","!sed -i 's\\\"\\\\g' manifest_tmp.tsv\n","!cat Metadata/manifest_tmp.tsv"],"metadata":{"id":"B2qqCBIHKJRM"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["#Creating the final manifest file.\n","#Change directory to the relevant folders and change pathnames accordingly.\n","\n","%%bash\n","cd Sequencing_data\n","for i in *.fastq.gz; do\n","    sed -i \"s|${i}|$PWD/${i}|g\" manifest_tmp.tsv;\n","done\n"],"metadata":{"id":"uCMrjrIVUPts"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["##Quality control using CutAdapt V4.2 and DADA2"],"metadata":{"id":"nwz6hYVYKdmK"}},{"cell_type":"markdown","source":["Cutadapt plugin in qiime2 was used to trim the set of primers from the paired end data."],"metadata":{"id":"t6bzqLj3ZVd3"}},{"cell_type":"code","source":["# Load qiime2 environment in the same directory as the data and open the jupyter notebook.\n","#Importing the demultiplexed paired end reads into qiime2.\n","!qiime tools import \\\n","  --type \"SampleData[PairedEndSequencesWithQuality]\" \\\n","  --input-format PairedEndFastqManifestPhred33V2 \\\n","  --input-path Sequencing_data/manifest_tmp.tsv \\\n","  --output-path demux_seqs.qza"],"metadata":{"id":"grK_nLmQDgOL"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["#Importing the cutadapt plugin in qiime2.\n","from qiime2.plugins import cutadapt"],"metadata":{"id":"VTtxQV7F--cV"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["#Using qiime2 plugin cutadapt to remove the primers.\n","!qiime cutadapt trim-paired \\\n","  --i-demultiplexed-sequences demultiplexed-seqs.qza \\\n","  --p-front-f CCTACGGGNGGCWGCAG \\\n","  --p-front-r GACTACHVGGGTATCTAATCC \\\n","  --p-error-rate 0 \\\n","  --o-trimmed-sequences trimmed-seqs.qza \\\n","  --verbose"],"metadata":{"id":"My8wBk179e0Y"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["DADA2 plugin for qiime2 was used for quality control. DADA2 helps with  filtering out noisy sequences, correcting errors in marginal sequences, removing chimeric sequences, removing singletons, joining denoised paired-end reads, and then dereplicating those sequences (https://docs.qiime2.org/2023.2/tutorials/overview/)."],"metadata":{"id":"yLE-ii-IZczD"}},{"cell_type":"markdown","source":["The parameter used below for the forward and reverse primers were chosen based on the paper from which the data was taken."],"metadata":{"id":"HxazJqdjZ7SQ"}},{"cell_type":"code","source":["#DADA2 was used to conduct quality control.\n","!qiime dada2 denoise-paired \\\n","  --i-demultiplexed-seqs trimmed-seqs.qza \\\n","  --p-trim-left-f 17 \\\n","  --p-trim-left-r 21 \\\n","  --p-trunc-len-f 280 \\\n","  --p-trunc-len-r 220 \\\n","  --p-trunc-q 2 \\\n","  --p-chimera-method  \"consensus\" \\\n","  --p-max-ee-f 2 \\\n","  --p-max-ee-r 2 \\\n","  --o-table space_table.qza \\\n","  --o-representative-sequences space_paired.qza \\\n","  --o-denoising-stats space_denoising_stats.qza"],"metadata":{"id":"oIxKkcdGMOZp"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["#Visualizing.\n","!qiime metadata tabulate \\\n","  --m-input-file space_denoising_stats.qza \\\n","  --o-visualization space_denoising_stats.qzv"],"metadata":{"id":"6jXO4fx8XZQd"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["##Taxonomy Classification: Training the Classifier from Greengenes2 using Naive Bayes in Qiime2"],"metadata":{"id":"CcSFLPSpXq9D"}},{"cell_type":"markdown","source":["First, the entire backbone sequence was obtained from Greengenes2 release."],"metadata":{"id":"9_SM4OlEYjBV"}},{"cell_type":"code","source":["!wget http://ftp.microbio.me/greengenes_release/2022.10/2022.10.backbone.full-length.fna.qza"],"metadata":{"id":"eqd4VqTJX71G"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["p-trunc-len was removed based on the suggestion by the tutorial about unequal lengths for the paired end sequences.\n","The lengths for 400 and 500 were chosen as per the forum chat and suggestion of a relaxed interval or allowed amplicon size to not miss biologically relevant data:https://forum.qiime2.org/t/picking-values-for-p-min-length-and-p-max-length-in-qiime-feature-classifier-extract-reads/20912."],"metadata":{"id":"zTy4JjhJY7LR"}},{"cell_type":"code","source":["#Extracting reads specific to the v3-v4 region.\n","!qiime feature-classifier extract-reads \\\n","  --i-sequences 2022.10.backbone.full-length.fna.qza \\\n","  --p-f-primer CCTACGGGNGGCWGCAG \\\n","  --p-r-primer GACTACHVGGGTATCTAATCC \\\n","  --p-min-length 400 \\\n","  --p-max-length 500 \\\n","  --o-reads gg2-ref-seqs.qza"],"metadata":{"id":"kdcmioUIYsjh"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["#Classifier was trained using Naiye Bayes\n","!qiime feature-classifier fit-classifier-naive-bayes \\\n","  --i-reference-reads gg2-ref-seqs.qza \\\n","  --i-reference-taxonomy 2022.10.backbone.tax.qza \\\n","  --o-classifier classifier.qza"],"metadata":{"id":"C2A_vqCeZEMY"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["##Predicting Taxonomy based on the Trained classifier"],"metadata":{"id":"3BuvcLihZfk-"}},{"cell_type":"markdown","source":["The classifier trained on the V3-V4 region: classifier.qza was used for taxonomy classification using qiime2 classify-sklearn."],"metadata":{"id":"5S9_3SuYZtL_"}},{"cell_type":"code","source":["!qiime feature-classifier classify-sklearn \\\n","  --i-classifier classifier.qza \\\n","  --i-reads space_paired.qza \\\n","  --o-classification space_taxonomy.qza"],"metadata":{"id":"I8PdpCwYZleb"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["##*De Novo* Phylogenetic Tree using IQTree"],"metadata":{"id":"RJc1eZiiabPJ"}},{"cell_type":"markdown","source":["To create a *De Novo* Phylogenetic tree, multiple sequence alignment of the ASVs is done using MAFFT and IQTree is used to create the most probable phylogenetic tree from that alignment. IQTree is based on maximum likelihood which means that it will find the tree most likely to be possible given the alignment from multiple possible trees."],"metadata":{"id":"HzMiyxKzYbmq"}},{"cell_type":"code","source":["!qiime phylogeny align-to-tree-mafft-iqtree \\\n","  --i-sequences space_paired.qza \\\n","  --o-alignment space_aligned.qza \\\n","  --o-masked-alignment space_masked_aligned.qza \\\n","  --o-tree space_unrooted_tree.qza \\\n","  --o-rooted-tree space_rooted_tree.qza \\\n","  --p-alrt 1000 \\\n","  --p-seed 42 \\\n","  --p-n-threads 8"],"metadata":{"id":"WyXjabw6U90i"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["The tree was visualized using Empress."],"metadata":{"id":"HsNaOa-uZBSy"}},{"cell_type":"code","source":["pip install empress"],"metadata":{"id":"IbOdhJbLVCD2"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["!qiime empress tree-plot \\\n","  --i-tree space_rooted_tree.qza \\\n","  --m-feature-metadata-file space_taxonomy.qza \\\n","  --o-visualization tree_viz.qzv"],"metadata":{"id":"q2ep8FLlVEGt"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["##Alpha Diversity: Shannon"],"metadata":{"id":"m2OTrFFAamlJ"}},{"cell_type":"markdown","source":["First the feature table is visualized to look at the minimum feature count and get a general distribution."],"metadata":{"id":"q4Go0L8Hw5PZ"}},{"cell_type":"code","source":["!qiime feature-table summarize \\\n","  --i-table space_table.qza \\\n","  --o-visualization space_table.qzv \\\n","  --m-sample-metadata-file space_tmp.tsv"],"metadata":{"id":"EER4XDgGw3_t"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Alpha rarefaction curves are generated at multiple sequencing depths and gives a visual of how the Shannon entropy changes at different depths for the three flight groups ad the control. All the curves plateau at ~500."],"metadata":{"id":"-vk-yfkXVbiZ"}},{"cell_type":"code","source":["#Creating the rarefaction curves with a maximum depth of 4700.\n","!qiime diversity alpha-rarefaction \\\n","  --i-table space_table.qza \\\n","  --i-phylogeny space_rooted_tree.qza \\\n","  --p-max-depth 4700 \\\n","  --m-metadata-file space_tmp.tsv \\\n","  --o-visualization space_shannon_alpha_rarefaction.qzv"],"metadata":{"id":"BEtlq_vJVUk8"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Looking at the feature table and the curves, a sampling depth of 4700 was chosen which excludes the controls from analyses and all other feature counts are above this count. Also, the curves are a plateau at this point. Rarefying at this depth would make all the feature counts 4700. "],"metadata":{"id":"FNQ5B7l2V1BA"}},{"cell_type":"code","source":["#Rarefying.\n","!qiime feature-table rarefy \\\n","  --i-table space_table.qza \\\n","  --p-sampling-depth 4700 \\\n","  --o-rarefied-table space_rarefied_table.qza"],"metadata":{"id":"kb5Des6JmIFJ"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Shannon metric is used for alpha diversity analysis which uses both evenness and richness of the community to give a measure of diversity within sample."],"metadata":{"id":"g7odY8h2WRqR"}},{"cell_type":"code","source":["#Plugin for alpha diversity.\n","!qiime diversity alpha \\\n","    --i-table space_rarefied_table.qza \\\n","    --p-metric shannon \\\n","    --o-alpha-diversity space_shannon_vector.qza"],"metadata":{"id":"FeI1yUQmVMpt"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["#Exporting and creating a tab separated file.\n","!qiime tools export --input-path space_shannon_vector.qza --output-path .\n","!mv alpha-diversity.tsv space_shannon.tsv"],"metadata":{"id":"-p3_MaE4VPcu"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Alpha diversity for each of the groups was calculated by grouping based on flight time. A Kruskal Wallis test is conduted to see if the different groups have different alpha diversity. A p value and H statistic is computed and pairwise comparisons for each of the groups are done with post hoc tests to look for difference between a pair of groups."],"metadata":{"id":"8fSpwYP1XSgE"}},{"cell_type":"code","source":["!qiime diversity alpha-group-significance \\\n","    --i-alpha-diversity space_shannon_vector.qza \\\n","    --m-metadata-file space_tmp.tsv \\\n","    --o-visualization space_shannon_group_significance.qzv"],"metadata":{"id":"tFeuOiuhVWuG"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["#Exporting and creating a tab separated file.\n","!qiime tools export --input-path space_shannon_vector.qza --output-path .\n","!mv alpha-diversity.tsv space_shannon.tsv"],"metadata":{"id":"GI34sK2DVYk2"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["##Beta Diversity using Weighted-UniFrac"],"metadata":{"id":"9Y6G8qZXvO4D"}},{"cell_type":"markdown","source":["To measure the difference in diversity between samples, beta diversity was measured. The metric used was Weighted-UniFrac that is phylogeny based and the branches of the phylogeny are abundance weighted."],"metadata":{"id":"IuLpH6doGnim"}},{"cell_type":"code","source":["#beta-phylogenetic to tell qiime2 to use the phylogeny based metrics. A distance matrix is created at the end.\n","!qiime diversity beta-phylogenetic \\\n","    --i-table space_rarefied_table.qza \\\n","    --i-phylogeny space_rooted_tree.qza \\\n","    --p-metric weighted_unifrac \\\n","    --o-distance-matrix space_weighted_unifrac_distance_matrix.qza"],"metadata":{"id":"oPcfKy-GhNRc"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["#Exporting the matrix and creating a tab separated file for the same.\n","!qiime tools export --input-path space_weighted_unifrac_distance_matrix.qza --output-path .\n","!mv distance-matrix.tsv space_weighted_unifrac.tsv"],"metadata":{"id":"J2fywupAhTia"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["#Visualizing the tsv file created.\n","!head space_weighted_unifrac.tsv"],"metadata":{"id":"Q8mYJre_hVyY"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["To look for differences based on flight time, Adonis test was conducted. Adonis test in qiime2 is wrapper of the adonis function in the vegan R package and therefore, cannot accept files with '-'. Therefore, files are first changed to the desirable format."],"metadata":{"id":"jsU41f8RN_o8"}},{"cell_type":"code","source":["#Removing '-' from the metadata file.\n","!cat space_tmp.tsv | tr -d '-' > space_tmp_rev.tsv"],"metadata":{"id":"SekWAySPha1j"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["!head space_tmp_rev.tsv"],"metadata":{"id":"kcC4ThCUhbl1"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["!cat space_tmp_rev.tsv | tr -d ' ' > space_tmp_rev_div.tsv"],"metadata":{"id":"Q1hzkILFhdeS"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["!head space_tmp_rev_div.tsv"],"metadata":{"id":"1FeoxnwohfUf"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["#Removing ':' from the headers.\n","!cat space_tmp_rev_div.tsv | tr -d ':' > space_tmp_rev.tsv"],"metadata":{"id":"0OCH6x7Ihhsi"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["!head space_tmp_rev.tsv"],"metadata":{"id":"DKgso27_hkM1"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["#Running the Adonis test using qiime2 plugin.\n","!qiime diversity adonis \\\n","  --i-distance-matrix space_weighted_unifrac_distance_matrix.qza \\\n","  --m-metadata-file space_tmp_rev.tsv \\\n","  --p-formula \"FactorValueTime\" \\\n","  --o-visualization weighted_unifrac_Time_adonis.qzv"],"metadata":{"id":"06kRXPUyhnUN"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["The Adonis test provides the F statistic which is the ratio of intergroup to intragroup variance and, in this case, is more than 1 but not large enough. The R2 value is the proportion of variance between the beta diversity of samples explained by the grouping variable in our case flight time which is ~5% and not very large."],"metadata":{"id":"HBI7HOqpPF7E"}},{"cell_type":"markdown","source":["##Principal Coordinate Analysis using Qiime2"],"metadata":{"id":"bneiTMaOFhp2"}},{"cell_type":"code","source":["!qiime diversity pcoa \\\n","    --i-distance-matrix space_weighted_unifrac_distance_matrix.qza \\\n","    --o-pcoa space_weighted_unifrac_pcoa_results.qza"],"metadata":{"id":"Q2cabRqMrBUG"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["EMPeror plugin in qiime2 was used to visualize the PCoA plot and look at the clutering pattern when colored by flight time."],"metadata":{"id":"X7tbHf0gFpiQ"}},{"cell_type":"code","source":["!qiime emperor plot \\\n","    --i-pcoa space_weighted_unifrac_pcoa_results.qza \\\n","    --m-metadata-file space_tmp_rev.tsv \\\n","    --o-visualization space_weighted_unifrac_emperor.qzv"],"metadata":{"id":"an80AAf4rG0J"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["A biplot was generated to look at the taxa that best explains the variance between flight groups in the PCoA generated before."],"metadata":{"id":"t502ZWDZF4Lc"}},{"cell_type":"code","source":["#The feature counts were normalized based on relative frequencies.\n","!qiime feature-table relative-frequency \\\n","  --i-table space_rarefied_table.qza \\\n","  --o-relative-frequency-table space_relative_table.qza"],"metadata":{"id":"trb-XX_GE-Gy"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["#Following command was used to overlay the taxa PCA over the sample PCoA.\n","!qiime diversity pcoa-biplot \\\n","  --i-pcoa space_weighted_unifrac_pcoa_results.qza \\\n","  --i-features space_relative_table.qza \\\n","  --o-biplot space_weighted_unifrac_pcoa_biplot.qza"],"metadata":{"id":"gvWGJiNiE_Cn"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["#Emperor was used to visualize the biplot.\n","!qiime emperor biplot \\\n","  --i-biplot space_weighted_unifrac_pcoa_biplot.qza \\\n","  --m-sample-metadata-file Metadata/space_tmp.tsv \\\n","  --m-feature-metadata-file space_taxonomy.qza \\\n","  --o-visualization space_weighted_unifrac_emperor_biplot.qzv"],"metadata":{"id":"5QWA0HSbFDOW"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["##Differential Abundance Analysis using ANCOM-BC"],"metadata":{"id":"yqcfcyD4vakc"}},{"cell_type":"markdown","source":["The taxa table was first collapsed to species, genus and family level (l7,l6 and l5 respectively)."],"metadata":{"id":"izvSzxKdwaN8"}},{"cell_type":"code","source":["#Collapsing the non-rarefied table to the specific taxa levels using a bash cell and for loop.\n","%%bash\n","for i in {5,6,7}; do\n","  qiime taxa collapse \\\n","    --i-table space_table.qza \\\n","    --i-taxonomy space_taxonomy.qza \\\n","    --p-level $i \\\n","    --o-collapsed-table l$i.qza\n","done"],"metadata":{"id":"B-X1RvtMvfxH"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["#Creating a visualization for each of the .qza files generated in the last cell.\n","%%bash\n","for i in {5,6,7}; do\n","  qiime feature-table summarize \\\n","    --i-table l$i.qza \\\n","    --o-visualization l$i.qzv \\\n","    --m-sample-metadata-file space_tmp.tsv\n","done\n","    "],"metadata":{"id":"Ydb6_1Hhv0XT"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["#Changing the way negative controls are labelled in the table.\n","!sed -i 's/notapplicable/control/g' space_tmp_rev.tsv"],"metadata":{"id":"x39Bqc0Lv4Xr"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["ANCOM-BC takes into consideration the compositionality of the microbiome and how changes in the sampling fraction used can change results of the abundance analysis. ANCOM-BC can only compare two groups at a time and therefore, a reference is chosen to be able to make comparisons between the three flight times. Since one of the questions of the study is to look whether during flight, the anxiety and depression levels, or such mental health issues are aggravated or more likely to occur due to space flight as a stressor, during flight is chosen as the reference for the ANCOM-BC comparison. ANCOM-BC will provide the log fold changes, p values and the q values. "],"metadata":{"id":"H4RWoX3m2pAw"}},{"cell_type":"code","source":["#Running ANCOM-BC on family level given the previous largely insignificant results for genus level with the diversity metrics.\n","!qiime composition ancombc \\\n","  --i-table l5.qza \\\n","  --m-metadata-file space_tmp_rev.tsv \\\n","  --p-reference-levels FactorValueTime::duringflight \\\n","  --p-formula FactorValueTime \\\n","  --o-differentials ancombc_space_time_during_flight_l5.qza"],"metadata":{"id":"Q25xJk4ZwGt6"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["#Creating a visualization of the ANCOM-BC results from the previous cell.\n","!qiime composition tabulate \\\n","  --i-data ancombc_space_time_during_flight_l5.qza \\\n","  --o-visualization ancombc_space_time_during_flight_l5.qzv"],"metadata":{"id":"KH7grBYjwUck"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["ANCOM-BC results can be visualized using another qiime2 plugin. To create this plot, the latest version of qiime2 (2023 version) was first installed. The visualization shows the taxa, at the family level, that are depleted or enriched with respect to during flight for both pre-flight and post-flight groups."],"metadata":{"id":"33GGQj3H3oNF"}},{"cell_type":"code","source":["#Creating the barplot.\n","!qiime composition da-barplot \\\n","  --i-data ancombc_space_time_during_flight_l5.qza \\\n","  --o-visualization ancombc_space_time_during_l5_barplot.qzv"],"metadata":{"id":"dWlUnq-XwVXI"},"execution_count":null,"outputs":[]}]}